# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: toolkit/components/remotebrowserutils/tests/browser/browser_httpCrossOriginOpenerPolicy.js
# Commit: 5686e9cc48b2
# Full Hash: 5686e9cc48b26a8ec550f94091d5b8b4ed840f7d
# Author: Valentin Gosu <valentin.gosu@gmail.com>
# Date: 2019-10-22 09:46:06
# Regressor Bug: 1579992
# File Overlap Count: 1
# Description:
#   Bug 1579992 - Load pages into new webCOOP+COEP process type r=nika
#   
#   * This patch makes pages with the `OPENER_POLICY_SAME_ORIGIN_EMBEDDER_POLICY_REQUIRE_CORP` policy load into a special `webCOOP+COEP={pageOrigin}` remote type.
#   * Adds `E10SUtils.WEB_REMOTE_COOP_COEP_TYPE_PREFIX="webCOOP+COEP="`
#   * When a COOP process switch occurs and the target page doesn't have this policy, we pass a `preferredRemoteType="web"` into `E10SUtils.getRemoteTypeForPrincipal` ensuring that we correctly get a different `remoteType`
# ==============================================================================

diff -r fd0d2f340380 -r 5686e9cc48b2 toolkit/components/remotebrowserutils/tests/browser/browser_httpCrossOriginOpenerPolicy.js
--- a/toolkit/components/remotebrowserutils/tests/browser/browser_httpCrossOriginOpenerPolicy.js	Mon Oct 21 16:34:43 2019 +0000
+++ b/toolkit/components/remotebrowserutils/tests/browser/browser_httpCrossOriginOpenerPolicy.js	Mon Oct 21 16:56:00 2019 +0000
@@ -25,7 +25,13 @@
   await loadedPromise;
 }
 
-async function test_coop(start, target, expectedProcessSwitch) {
+async function test_coop(
+  start,
+  target,
+  expectedProcessSwitch,
+  startRemoteTypeCheck,
+  targetRemoteTypeCheck
+) {
   return BrowserTestUtils.withNewTab(
     {
       gBrowser,
@@ -37,11 +43,16 @@
 
       await new Promise(resolve => setTimeout(resolve, 20));
       let browser = gBrowser.selectedBrowser;
-      let firstProcessID = await ContentTask.spawn(browser, null, () => {
-        return Services.appinfo.processID;
-      });
+      let firstRemoteType = browser.remoteType;
+      let firstProcessID = browser.frameLoader.remoteTab.osPid;
 
-      info(`firstProcessID: ${firstProcessID}`);
+      info(
+        `firstProcessID: ${firstProcessID} firstRemoteType: ${firstRemoteType}`
+      );
+
+      if (startRemoteTypeCheck) {
+        startRemoteTypeCheck(firstRemoteType);
+      }
 
       await performLoad(
         browser,
@@ -63,11 +74,15 @@
       info(`Navigated to: ${target}`);
       await new Promise(resolve => setTimeout(resolve, 20));
       browser = gBrowser.selectedBrowser;
-      let secondProcessID = await ContentTask.spawn(browser, null, () => {
-        return Services.appinfo.processID;
-      });
+      let secondRemoteType = browser.remoteType;
+      let secondProcessID = browser.frameLoader.remoteTab.osPid;
 
-      info(`secondProcessID: ${secondProcessID}`);
+      info(
+        `secondProcessID: ${secondProcessID} secondRemoteType: ${secondRemoteType}`
+      );
+      if (targetRemoteTypeCheck) {
+        targetRemoteTypeCheck(secondRemoteType);
+      }
       if (expectedProcessSwitch) {
         Assert.notEqual(
           firstProcessID,
@@ -118,7 +133,7 @@
     info(`test_download: Test tab ready`);
 
     let start = httpURL(
-      "coop_header.sjs?downloadPage&" + initCoop,
+      "coop_header.sjs?downloadPage&coop=" + initCoop,
       "https://example.com"
     );
     await performLoad(
@@ -195,7 +210,10 @@
       Assert.equal(prevPID, currentPID);
       prevPID = currentPID;
 
-      target = httpURL("coop_header.sjs?same-origin", "https://example.org");
+      target = httpURL(
+        "coop_header.sjs?coop=same-origin",
+        "https://example.org"
+      );
       await performLoad(
         browser,
         {
@@ -219,7 +237,10 @@
       Assert.notEqual(prevPID, currentPID);
       prevPID = currentPID;
 
-      target = httpURL("coop_header.sjs?same-origin", "https://example.com");
+      target = httpURL(
+        "coop_header.sjs?coop=same-origin",
+        "https://example.com"
+      );
       await performLoad(
         browser,
         {
@@ -243,7 +264,10 @@
       Assert.notEqual(prevPID, currentPID);
       prevPID = currentPID;
 
-      target = httpURL("coop_header.sjs?same-origin.#4", "https://example.com");
+      target = httpURL(
+        "coop_header.sjs?coop=same-origin&index=4",
+        "https://example.com"
+      );
       await performLoad(
         browser,
         {
@@ -274,43 +298,120 @@
     false
   );
   await test_coop(
-    httpURL("coop_header.sjs?same-origin", "http://example.com"),
+    httpURL("coop_header.sjs?coop=same-origin", "http://example.com"),
     httpURL("coop_header.sjs", "http://example.com"),
     false
   );
   await test_coop(
     httpURL("coop_header.sjs", "http://example.com"),
-    httpURL("coop_header.sjs?same-origin", "http://example.com"),
+    httpURL("coop_header.sjs?coop=same-origin", "http://example.com"),
     false
   );
   await test_coop(
-    httpURL("coop_header.sjs?same-origin", "http://example.com"),
-    httpURL("coop_header.sjs?same-site", "http://example.com"),
+    httpURL("coop_header.sjs?coop=same-origin", "http://example.com"),
+    httpURL("coop_header.sjs?coop=same-site", "http://example.com"),
     false
   ); // assuming we don't have fission yet :)
 });
 
 add_task(async function test_enabled() {
   await SpecialPowers.pushPrefEnv({ set: [[PREF_NAME, true]] });
+
+  function checkIsCoopRemoteType(remoteType) {
+    Assert.ok(
+      remoteType.startsWith(E10SUtils.WEB_REMOTE_COOP_COEP_TYPE_PREFIX),
+      `${remoteType} expected to be coop`
+    );
+  }
+
+  function checkIsNotCoopRemoteType(remoteType) {
+    if (gFissionBrowser) {
+      Assert.ok(
+        remoteType.startsWith("webIsolated="),
+        `${remoteType} expected to start with webIsolated=`
+      );
+    } else {
+      Assert.equal(
+        remoteType,
+        E10SUtils.WEB_REMOTE_TYPE,
+        `${remoteType} expected to be web`
+      );
+    }
+  }
+
   await test_coop(
     httpURL("coop_header.sjs", "https://example.com"),
     httpURL("coop_header.sjs", "https://example.com"),
-    false
+    false,
+    checkIsNotCoopRemoteType,
+    checkIsNotCoopRemoteType
   );
   await test_coop(
     httpURL("coop_header.sjs", "https://example.com"),
-    httpURL("coop_header.sjs?same-origin", "https://example.org"),
-    true
+    httpURL("coop_header.sjs?coop=same-origin", "https://example.org"),
+    true,
+    checkIsNotCoopRemoteType,
+    checkIsNotCoopRemoteType
+  );
+  await test_coop(
+    httpURL("coop_header.sjs?coop=same-origin&index=1", "https://example.com"),
+    httpURL("coop_header.sjs?coop=same-origin&index=1", "https://example.org"),
+    true,
+    checkIsNotCoopRemoteType,
+    checkIsNotCoopRemoteType
+  );
+  await test_coop(
+    httpURL("coop_header.sjs?coop=same-origin&index=2", "https://example.com"),
+    httpURL("coop_header.sjs?coop=same-site&index=2", "https://example.org"),
+    true,
+    checkIsNotCoopRemoteType,
+    checkIsNotCoopRemoteType
+  );
+  await test_coop(
+    httpURL("coop_header.sjs", "https://example.com"),
+    httpURL(
+      "coop_header.sjs?coop=same-origin&coep=require-corp",
+      "https://example.com"
+    ),
+    true,
+    checkIsNotCoopRemoteType,
+    checkIsCoopRemoteType
   );
   await test_coop(
-    httpURL("coop_header.sjs?same-origin#1", "https://example.com"),
-    httpURL("coop_header.sjs?same-origin#1", "https://example.org"),
-    true
+    httpURL(
+      "coop_header.sjs?coop=same-origin&coep=require-corp&index=2",
+      "https://example.com"
+    ),
+    httpURL(
+      "coop_header.sjs?coop=same-origin&coep=require-corp&index=3",
+      "https://example.com"
+    ),
+    false,
+    checkIsCoopRemoteType,
+    checkIsCoopRemoteType
   );
   await test_coop(
-    httpURL("coop_header.sjs?same-origin#2", "https://example.com"),
-    httpURL("coop_header.sjs?same-site#2", "https://example.org"),
-    true
+    httpURL(
+      "coop_header.sjs?coop=same-origin&coep=require-corp&index=4",
+      "https://example.com"
+    ),
+    httpURL("coop_header.sjs", "https://example.com"),
+    true,
+    checkIsCoopRemoteType,
+    checkIsNotCoopRemoteType
+  );
+  await test_coop(
+    httpURL(
+      "coop_header.sjs?coop=same-origin&coep=require-corp&index=5",
+      "https://example.com"
+    ),
+    httpURL(
+      "coop_header.sjs?coop=same-origin&coep=require-corp&index=6",
+      "https://example.org"
+    ),
+    true,
+    checkIsCoopRemoteType,
+    checkIsCoopRemoteType
   );
 });
 